libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,200 @@
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/analysis/FunctionCallGraph.h>
+#include <boost/range/adaptor/reversed.hpp>
+
+using namespace std;
+using namespace solidity::frontend;
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(Node const& _lhs, Node const& _rhs) const
+{
+	if (_lhs.index() != _rhs.index())
+		return _lhs.index() < _rhs.index();
+
+	if (std::holds_alternative<SpecialNode>(_lhs))
+		return std::get<SpecialNode>(_lhs) < std::get<SpecialNode>(_rhs);
+	return std::get<ASTNode const*>(_lhs)->id() < std::get<ASTNode const*>(_rhs)->id();
+}
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(Node const& _lhs, int64_t _rhs) const
+{
+	solAssert(!std::holds_alternative<SpecialNode>(_lhs), "");
+
+	return std::get<ASTNode const*>(_lhs)->id() < _rhs;
+}
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(int64_t _lhs, Node const& _rhs) const
+{
+	solAssert(!std::holds_alternative<SpecialNode>(_rhs), "");
+
+	return _lhs < std::get<ASTNode const*>(_rhs)->id();
+}
+
+unique_ptr<FunctionCallGraphBuilder::ContractCallGraph> FunctionCallGraphBuilder::create(ContractDefinition const& _contract)
+{
+	FunctionCallGraphBuilder builder(_contract);
+
+	// Create graph for constructor, state vars, etc
+	builder.m_currentNode = SpecialNode::EntryCreation;
+	builder.m_currentDispatch = SpecialNode::InternalCreationDispatch;
+
+	for (ContractDefinition const* contract: _contract.annotation().linearizedBaseContracts | boost::adaptors::reversed)
+	{
+		for (auto const* stateVar: contract->stateVariables())
+			stateVar->accept(builder);
+
+		for (auto arg: contract->baseContracts())
+			arg->accept(builder);
+
+		if (contract->constructor())
+		{
+			builder.add(*builder.m_currentNode, contract->constructor());
+			contract->constructor()->accept(builder);
+			builder.m_currentNode = contract->constructor();
+		}
+	}
+
+	builder.m_currentNode.reset();
+	builder.m_currentDispatch = SpecialNode::InternalDispatch;
+
+	// Create graph for all publicly reachable functions
+	for (auto& [hash, functionType]: _contract.interfaceFunctionList())
+	{
+		(void)hash;
+		if (auto const* funcDef = dynamic_cast<FunctionDefinition const*>(&functionType->declaration()))
+			if (!builder.m_graph->edges.count(funcDef))
+				builder.visitCallable(funcDef);
+
+		// Add all external functions to the RuntimeDispatch
+		builder.add(SpecialNode::Entry, &functionType->declaration());
+	}
+
+	// Add all InternalCreationDispatch calls to the RuntimeDispatch as well
```

Comment:
Because that's what the codegen backend does as well, from what I understood of what chris told me

Reply:
I'm not sure in what way codegen would do this... this might rather hide the other half of the issues due to https://github.com/ethereum/solidity/pull/10332#discussion_r553509879 - so it may be worth checking what happens if this is removed after that one is fixed.

Reply:
This at least needs a comment explaining.

Reply:
I think the idea was as follows: A function is added to the creation dispatch if it is accessed without being called right away. We don't know what happens with that function variable afterwards - it might be stored in storage. If it is stored in storage, it might be called after deploy time. Because of that, code that calls the deployed dispatch might also call any function added to the creation dispatch.

Reply:
Yeah, that makes sense.

Reply:
The comment is about: A function referenced (but not called) at creation time has to be included in the internal dispatch after deployment, because it could have been stored in storage. Essentially, this means that all outgoing nodes of the internal dispatch at creation time are also outgoing nodes of the internal dispatch at post-deploy time. There is one difference, though: They do not go to the same nodes, they just go to the same functions. So the best way would probably be to just have a set / list of functions that are referenced (but not call) at creation time.

Reply:
I meant that the comment in the snippet visible above says `RuntimeDispatch` while the line below deals with internal dispatch. I'll change it.

Anyway, looks like the code used to add all the functions to both dispatches but now there's just a single `add(SpecialNode::InternalDispatch, SpecialNode::InternalCreationDispatch)` there so I think this is resolved.